### PR TITLE
Fix Kubernetes endpoints subsets behavior

### PR DESCRIPTION
**Pull request by tgraf**
The Kubernetes Endpoints resource is split into a list of subsets, each consists of a list of addresses and a list of ports. So far, all subsets have been merged together, creating the cross product over addresses and ports of all subsets. This is incorrect. Each subset must have its cross product created before being merged together to a list of backends.

This is a dependency for clustermesh service discovery

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6225)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-missed-k8s

```
	 default       kafka-broker-79f447bc95-x6dn9          0/1       ImagePullBackOff   0          5m        10.10.0.106     k8s1
```

**Comment by tgraf** [comment]
test-missed-k8s

connectivity test during upgrade

**Comment by tgraf** [comment]
test-missed-k8s

**Review by ianvernon — APPROVED** [review]

File: pkg/k8s/endpoints.go [hunk 1]
```diff
@@ -32,43 +32,28 @@ import (
...
-
-	ports := make([]string, len(e.Ports))
-	i = 0
-	for p := range e.Ports {
-		ports[i] = string(p)
-		i++
+	backends := []string{}
+	for ip, ports := range e.Backends {
+		for _, port := range ports {
+			backends = append(backends, fmt.Sprintf("%s:%d/%s", ip, port.Port, port.Protocol))
```
Comment (aanm) [inline pkg/k8s/endpoints.go]: `backends = append(backends, fmt.Sprintf("%s/%s", net.JoinHostPort(ip, port.Port), port.Protocol))`

File: pkg/k8s/endpoints.go [hunk 2]
```diff
@@ -32,43 +32,28 @@ import (
 // (Ports). Each k8s endpoint present in BEIPs share the same list of Ports
 // open.
 type Endpoints struct {
-	// TODO: Replace bool for time.Time so we know last time the service endpoint was seen?
-	BackendIPs map[string]bool
-	Ports      map[loadbalancer.FEPortName]*loadbalancer.L4Addr
+	Backends map[string]service.PortConfiguration
```
Comment (aanm) [inline pkg/k8s/endpoints.go:39]: Write comment what is the key and to what does it map to

File: pkg/k8s/endpoints.go [hunk 3]
```diff
@@ -81,29 +66,39 @@ func (e *Endpoints) DeepEquals(o *Endpoints) bool {
...
+	for ip1, ports1 := range e.Backends {
+		ports2, ok := o.Backends[ip1]
+		if !ok {
+			return false
+		}
 
-	for k1, v1 := range e.Ports {
-		v2, ok := o.Ports[k1]
-		if !ok || !v1.Equals(v2) {
+		if len(ports1) != len(ports2) {
```
Comment (aanm) [inline pkg/k8s/endpoints.go]: Should be a DeepEquals implementation of  `service.PortConfiguration`

File: pkg/k8s/endpoints.go [hunk 4]
```diff
@@ -81,29 +66,39 @@ func (e *Endpoints) DeepEquals(o *Endpoints) bool {
...
 
 	return true
 }
 
 // CIDRPrefixes returns the endpoint's backends as a slice of IPNets.
 func (e *Endpoints) CIDRPrefixes() ([]*net.IPNet, error) {
-	prefixes := make([]string, 0, len(e.BackendIPs))
-	for backend := range e.BackendIPs {
-		prefixes = append(prefixes, backend)
+	prefixes := make([]string, len(e.Backends))
```
Comment (aanm) [inline pkg/k8s/endpoints.go:93]: Just a small note, go compiler was optimized so between this change and the way it was before there's no major difference in terms of performance.

**Review by aanm — DISMISSED** [review]

**Comment by aanm** [comment]
Should we do backport of this? :thinking:

**Comment by tgraf** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**